finality-aleph/src/base_protocol/handler.rs

```diff
@@ -1,45 +1,147 @@
+            TooManyPeers(role, direction) => {
+                write!(f, "too many {:?} nodes connected {:?}", role, direction)
+            }
+        }
+    }
 }
 
-/// Handler for the base protocol.
+struct ConnectionLimiter {
+    num_full_in_peers: usize,
+    num_full_out_peers: usize,
+    num_light_peers: usize,
+    max_full_in_peers: usize,
+    max_full_out_peers: usize,
+    max_light_peers: usize,
+}
+
+impl ConnectionLimiter {
+    pub fn new(net_config: &NetworkConfiguration) -> Self {
+        // It is assumed that `default_peers_set.out_peers` only refers to full nodes, but
+        // `default_peers_set.in_peers` refers to both full and light nodes.
+        // Moreover, `default_peers_set_num_full` refers to the total of full nodes.
+        let max_full_out_peers = net_config.default_peers_set.out_peers as usize;
+        let max_full_in_peers =
+            (net_config.default_peers_set_num_full as usize).saturating_sub(max_full_out_peers);
+        let max_light_peers =
+            (net_config.default_peers_set.in_peers as usize).saturating_sub(max_full_in_peers);
+        ConnectionLimiter {
+            num_full_in_peers: 0,
+            num_full_out_peers: 0,
+            num_light_peers: 0,
+            max_full_in_peers,
+            max_full_out_peers,
+            max_light_peers,
+        }
+    }
+
+    pub fn allowed(&self, peer: &PeerInfo) -> bool {
+        match (peer.role, peer.direction) {
+            (Role::Light, _) => self.num_light_peers < self.max_light_peers,
+            (Role::Full, Direction::Inbound) => self.num_full_in_peers < self.max_full_in_peers,
+            (Role::Full, Direction::Outbound) => self.num_full_out_peers < self.max_full_out_peers,
+        }
+    }
+
+    fn count_for(&mut self, peer: &PeerInfo) -> &mut usize {
+        match (peer.role, peer.direction) {
+            (Role::Light, _) => &mut self.num_light_peers,
+            (Role::Full, Direction::Inbound) => &mut self.num_full_in_peers,
+            (Role::Full, Direction::Outbound) => &mut self.num_full_out_peers,
+        }
+    }
+
+    pub fn add(&mut self, peer: &PeerInfo) {
+        self.count_for(peer).saturating_inc();
+    }
+
+    pub fn remove(&mut self, peer: &PeerInfo) {
+        self.count_for(peer).saturating_dec();
+    }
+}
+
+/// Handler for the base protocol. Deals with accepting and counting connections.
 pub struct Handler<B>
 where
     B: Block<Hash = BlockHash>,
     B::Header: Header<Number = BlockNumber>,
 {
     reserved_nodes: HashSet<PeerId>,
     peers: HashMap<PeerId, PeerInfo>,
-    // the below counters and bounds ignore the nodes which belong to `reserved_nodes`
-    num_full_in_peers: usize,
-    num_full_out_peers: usize,
-    num_light_peers: usize,
-    max_full_in_peers: usize,
-    max_full_out_peers: usize,
-    max_light_peers: usize,
+    // the limiter ignores the nodes which belong to `reserved_nodes`
+    limiter: ConnectionLimiter,
```

Comment:
```suggestion
    connection_limits: ConnectionLimits,
```

Reply:
Done, although only "limits" as it should be enough.